src/public/javascripts/State.js

```diff
@@ -0,0 +1,144 @@
+let StateButton = require('../assets/StateButton');
+const CLIENT = require('../javascripts/communication');
+const DYNAMIC_LOADING = require('../javascripts/dynamicloading');
+
+let STATES = [];
+/**
+ * Class reperesenting a state on the state machine
+ */
+class State {
+  /**
+   * Creates a state
+   * @param {String} shortName code name for the state, no spaces
+   * @param {String} displayName Display name for the buttons and console logs, any format
+   * @param {Button} btn (Optional) Button to assign to state if button is already made,
+   * if creating a button assign null
+   * @param {String} btnColor Color to create state button
+   * @param {Boolean} isHazardus If the state requires confirmation to transition to
+   * @param {Boolean} isFault If the state is a fault state
+   */
+  constructor(shortName, displayName, btn, btnColor, isHazardus, isFault) {
+    this.shortname = shortName;
+    this.displayName = displayName;
+    this.isHazardus = isHazardus;
+    this.idNum = STATES.length;
+    this.isFault = isFault;
+    if (btn) {
+      this.btn = btn;
+    } else {
+      this.btn = new StateButton(shortName, displayName, null, btnColor, isHazardus, this);
+    }
+    this.active = false;
+    STATES.push(this);
+  }
+
+  /**
+   * Either prompts the user to confirm activation or sets active state
+   * @param {HTMLElement} modalTemplate The template of the modal to display
+   */
+  activate(modalTemplate, fromPod) {
+    if (this.isHazardus && !fromPod) {
+      this.confirmActive(modalTemplate);
+    } else {
+      this.setActive();
+    }
+  }
+
+  /**
+   * Sets the state as the active state
+   */
+  setActive() {
+    this.active = true;
+    STATES.forEach((state) => {
+      state.setInactive();
+    });
+    this.active = true;
+    this.btn.activate();
+    CLIENT.sendOverride(this.shortName);
+    if (!this.isFault) {
+      DYNAMIC_LOADING.fillAllBounds(this.shortname);
+    }
+  }
+
+  /**
+   * Sets the state as an inactive state
+   */
+  setInactive() {
+    this.active = false;
+    this.btn.deactivate();
+  }
+
+  /**
+   * Displays a confirmation window before transitioning to state
+   * @param {HTMLElement} modalTemplate The template of the modal to display
+   * @override
+   */
+  confirmActive(modalTemplate) {
+    let modal = modalTemplate.cloneNode(true);
+    document.body.appendChild(modal);
+    State.toggleConfirmationModal(modal);
+    let button = modal.childNodes[1].childNodes[7];
+    let text = modal.childNodes[1].childNodes[5];
+    let closeButton = modal.childNodes[1].childNodes[1];
+    closeButton.addEventListener('click', () => {

```

Comment:
Same as before, won't this stack event listeners?

Reply:
Since there is no (intuitive) way to remove an event listener from an object what this is doing is creating an entirely new modal from a template. That way every time `confirmActive()` is called the buttons have no event listeners and I can set them entirely new depending on what button was pressed before it.